src/lib/communication/DY.Lib.Communication.API.Lemmas.fst

```diff
@@ -0,0 +1,312 @@
+      let (msg_id, tr) = send_msg msg_signed tr in
+      assert(tr_out == tr);
+      ()
+    )
+  
+
+#push-options "--ifuel 1 --fuel 0"
+val verify_message_proof:
+  {|cinvs:crypto_invariants|} ->
+  tr:trace ->
+  sender:principal -> receiver:principal ->
+  msg_bytes:bytes -> vk_sender:bytes ->
+  Lemma
+  (requires
+    has_communication_layer_invariants cinvs /\
+    is_publishable tr msg_bytes /\
+    is_verification_key (SigKey comm_layer_sign_tag empty) (principal_label sender) tr vk_sender
+  )
+  (ensures (
+    match verify_message receiver msg_bytes vk_sender with
+    | Some {sender=sender'; receiver=receiver'; payload} -> (
+      is_publishable tr payload /\
+      ((get_signkey_label vk_sender == principal_label sender' /\
+        sender' == sender /\
+        event_triggered tr sender' (CommAuthSendMsg sender' payload)) \/
+        get_signkey_label vk_sender `can_flow tr` public)      
+    )
+    | _ -> True
+  ))
+let verify_message_proof #cinvs tr sender receiver msg_bytes vk_sender =
+  assert(get_signkey_label vk_sender == principal_label sender);
+  match verify_message receiver msg_bytes vk_sender with
+  | Some {sender=sender'; receiver=receiver'; payload} -> (
+    parse_wf_lemma communication_message (is_publishable tr) msg_bytes;
+    let msg = CM {sender=sender'; receiver=receiver'; payload} in
+    serialize_wf_lemma communication_message #parseable_serializeable_bytes_communication_message (is_publishable tr) msg;
+    ()
+  )
+  | _ -> ()
+#pop-options
+
+val extract_data_from_msg_id:
+  tr:trace ->
+  comm_keys_ids:communication_keys_sess_ids ->
+  receiver:principal -> msg_id:timestamp ->
+  (principal -> principal -> bytes -> prop) ->
+  prop
+let extract_data_from_msg_id tr comm_keys_ids receiver msg_id fn =
+  match recv_msg msg_id tr with
+  | (None, tr) -> True
+  | (Some msg_signed, tr) -> (
+    match get_sender msg_signed with
+    | None -> True
+    | Some sender -> (
+      match get_public_key receiver comm_keys_ids.pki (Verify comm_layer_sign_tag) sender tr with
+      | (None, tr) -> True
+      | (Some vk_sender, tr) -> (
+        match verify_message receiver msg_signed vk_sender with
+        | None -> True
+        | Some {sender=sender'; receiver=receiver'; payload} -> (
+          fn sender receiver payload
+        )
+      )
+    )
+  )
+
+val receive_authenticated_proof:
+  {|invs:protocol_invariants|} ->
+  tr:trace ->
+  higher_layer_preds:comm_higher_layer_event_preds ->
+  comm_keys_ids:communication_keys_sess_ids ->
+  receiver:principal -> msg_id:timestamp ->
+  Lemma
+  (requires
+    trace_invariant tr /\ has_pki_invariant invs /\
+    has_communication_layer_invariants invs.crypto_invs /\
+    (
+      extract_data_from_msg_id tr comm_keys_ids receiver msg_id (
+        fun sender receiver payload -> (
+          comm_higher_layer_event_preds_later tr (
+            higher_layer_preds.receive_auth sender receiver payload
+          )
+        )
+      )
+    ) /\
```

Comment:
How are we supposed to prove this precondition?

Reply:
This is also removed now.